Remove dependency on BotContext for IntentRecognizerMiddleware. Add fluent intent handers.

garypretty opened the pull request:
@cleemullins Raised this PR as its the easiest way to get your thoughts. I have been thinking about how we could implement a useful generic middleware for Intent handling, with the comments you made around not having the TopIntent information on the BotContext. 

**More work to do, but,** I have updated IntentRecognizerMiddleware to support fluent style intent handlers, where the user can add the middleware using .Use() but then assign intent handlers using .HandleIntent(, similar to the below.  We could obviously add options to end / continue routing and more etc...

```cs
            if (adapter == null)
            {
                adapter = new BotFrameworkAdapter(configuration)
                    .Use(new LuisRecognizerMiddleware("xxxxxx", "xxxxxx"))
                    .HandleIntent("MyFirstIntent", HandleMyFirstIntent)
                    .HandleIntent("MySecondIntent", HandleMySecondIntent));
```

with an intent handler that looks something like....

```cs
        private Task HandleMyFirstIntent(IBotContext context, Intent intent)
        {
            context.Reply($"You triggered the intent {intent.Name} intent with a score of {intent.Score}");
            return Task.CompletedTask;
        }
```

I implemented a basic version of this and I really like it, especially as it keeps the whole thing very isolated and makes the understanding of what will happen when an intent is triggered really easy to understand (it actually maps very nicely to the way developers handle intents right now in v3 with a method decorated with the intent name).

Obviously, the best thing is the fact that we can remove the need to have anything to do with Intents on the BotContext at all.

Would be great to get your thoughts.

cleemullins commented:
@Stevenic - your thoughts?

garypretty commented:
@cleemullins @johnataylor Obviously if this is an approach we move forward, or a derivative of it, then I will tidy / tighten up the code and update all of the necessary tests etc.

JonathanFingold commented:
@garypretty How would this work with multiple recognizers?
1. Run two recognizers in sequence, if not satisfied with the result of the first, "fail-over" to the second.
1. Run two in parallel, choose the "best" result of the two.

I think I've heard that both scenarios are currently in use for v3.

garypretty commented:
The ability to chain the middleware together is not a problem. On the PR right now routing to additional middleware is stopped, but this was simply to get initial thoughts before I proceed in making things more robust. So giving the option to failover to another instance of the same middleware or another type, such as QnA maker is no problem. 

The second point about running them in parallel and running the one with the highest score would need a little more thinking about. You are right that v3 supports these options right now and I think there are be a few ways we could do it, but I think that particular use case is secondary and something that we can consider as an enhancements.

garypretty commented:
@cleemullins @Stevenic I have updated this PR with an updated to the sample for Luis with comments and showing an example usage of the handlers and also added the option to end activity routing on successful handling of an intent (true by default).

garypretty commented:
@Stevenic, did you get s chance to take a look at this? Be great to get your thoughts.

cleemullins commented:
Hey Gary,

I’ll go through this tonight. To some degree, it’s similar to tom’s branch that is currently in Review. In there he has revisited the entire Context model in a more robust way.

garypretty commented:
@cleemullins yes, I agree. I saw that branch. In my head at the moment I can almost see a hybrid approach. I.e. The approach Tom has taken, with the added ability to have intent handlers (as in my branch) as an option on the middleware. If not intent handlers are used then the results still get placed on the context regardless, but I still think having the handlers option is worthwhile.

garypretty commented:
@cleemullins did you get a chance to think through this last night? No worries if not, I know you are busy. Just wondered :)

cleemullins commented:
I **think** what we're going to be doing in the near term is to focus on the next iteration of the Middleware pipeline. This is a "Single Pipeline" (instead of the 3+2 model we have now). Steve had been blazing the trail here in the Node Repo in the [M2 branch](https://github.com/Microsoft/botbuilder-js/tree/m2/libraries). 

Specifically, if you look at the [Core library in the M2 Branch](https://github.com/Microsoft/botbuilder-js/tree/m2/libraries/botbuilder-core) you can see: 

1. Updated ["Lightweight" context](https://github.com/Microsoft/botbuilder-js/blob/m2/libraries/botbuilder-core/src/botContext.ts). 
2. The updated [BotAdapter](https://github.com/Microsoft/botbuilder-js/blob/m2/libraries/botbuilder-core/src/botAdapter.ts), with abstract methods for the receive handlers. 
3. Some further simplification. 

There is discussion threads around all of this, and the simplified programming model seems right. It also better aligns with Restify, Express, ASP.Net Core, and Redux models of middleware. Once of the best features of this is that object lifetime is easier to manage. In today's C# code, the (for example) ContextCreated pipeline is good for creation, but there are no clear indicators when to tear things down. This awkward lifecycle management will become a problem as things get more complex. 


We anticipate this is the final "Large" change to the core patterns we're using.

garypretty commented:
Ok @cleemullins that makes sense. We can revisit this after that change. In the meantime I know your age going to setup a toybox repo for. Net. Did you do this? I think it makes sense to have one for the common pieces of middleware etc and I could put a derivative of this in there for now.

tomlm commented:
Coming to this party late.  It's interesting, because we have been flirting with intent handlers as a registered piece of middleware, but I think we have been moving away from that and instead thinking of the middleware as truly things that are global.  The way I would restructure what you are doing now would be to have a dispatch class which does your intent dispatching inside of the OnReceive() of your bot.

One line I have been drawing in my head is to think, "is this bot logic, or something that is bot agnostic?"  A storage provider is bot agnostic, translation, logging are all bot agnostic and make sense as middleware.  A intent matcher with dispatch handlers seems like a layer on top of that.

```c#
var adapter = new Adapter().Use(ConversationState<MyState>());

adapter.ProcessActivity((ctx) =>
{
     IntentHandlers route = new IntentHandlers()
              .AddIntent("xxxx", myHandler1)
              .AddIntent("yyy", myHandler2)
);
```
I don't know, what do you think?

garypretty commented:
Hey @tomlm. Thanks for chiming in on this :)  I like the idea of the dispatch class and totally see where you are going with it - it could definitely help.  However, I do think that, for LUIS especially, giving the developers the choice of using the intent handlers could be useful.  So, I guess I am suggesting that we offer both the current functionality of having the intent matching info set on the context AND the option of intent handlers.  e.g. You use the middleware - if you don't specify any intent handlers then it works as it does now, but we add the handler support and the handler kicks in if you have added one for the top matching intent.

I think this gives flexibility in a couple of ways.  The first is that this pattern will be very familiar for many v3 devs who have used the LUIS dialog. Secondly though I think that having the handlers on the intent can give some valued flexibility when it comes to middleware chaining.  One of the major use cases that I have used (and I know many devs I talk to in the community use) is the ability to use LUIS to handle intents, but then fall back to something like QnA Maker if no intent matches.  By having the handlers we can give the developer the ability to handle and intent and then, if they want to, they can prevent further routing to additional middleware - but if no intent matches then we can fall into the QnA Maker Middleware.

Given that this is fairly straightforward for us to implement I think it makes sense and then we could get feedback from people over the coming weeks / months.  My proposal therefore would be that given we have PR #228 bringing a redeveloped LUIS recognizer to the party, I wait until that is merged and then add the handler support to that.

I know that often having a single way of achieving something brings simplicity and clarity, but I think in this case the handler support would be very beneficial. 

What do you think?

cleemullins commented:
@garypretty 

> My proposal therefore would be that given we have PR #228 bringing a redeveloped LUIS middleware and recognizer to the party, I wait until that is merged and then add the handler support to that.

Sounds good. We're just waiting on a new [Nuget package being published here](https://www.nuget.org/packages/Microsoft.Cognitive.LUIS/) before we can take that LUIS PR. That's moving forward, but has hit minor logistical complications. We *think* we have resolved everything, and should hopefully be ready to merge soon. @emadelwany - When #228 is ready, we've got more good stuff queued up!

garypretty commented:
Excellent. I'll wait until that's merged and then raise a new PR with my changes for handlers for review. Btw @cleemullins and team. I'm flying out to you guys first thing tomorrow. Look forward to meeting you guys in person.

cleemullins commented:
Closing per @garypretty 's last comment
